src/auth/oauth2/qgso2.cpp

```diff
@@ -348,3 +350,126 @@ QNetworkAccessManager *QgsO2::getManager()
 {
   return QgsNetworkAccessManager::instance();
 }
+
+/// Parse JSON data into a QVariantMap
+static QVariantMap parseTokenResponse( const QByteArray &data )
+{
+  QJsonParseError err;
+  QJsonDocument doc = QJsonDocument::fromJson( data, &err );
+  if ( err.error != QJsonParseError::NoError )
+  {
+    qWarning() << "parseTokenResponse: Failed to parse token response due to err:" << err.errorString();
+    return QVariantMap();
+  }
+
+  if ( !doc.isObject() )
+  {
+    qWarning() << "parseTokenResponse: Token response is not an object";
+    return QVariantMap();
+  }
+
+  return doc.object().toVariantMap();
+}
+
+// Code adapted from O2::refresh() with the addition of an event loop
+void QgsO2::refreshSynchronous()
+{
+  qDebug() << "O2::refresh: Token: ..." << refreshToken().right( 7 );
+
+  if ( refreshToken().isEmpty() )
+  {
+    qWarning() << "O2::refresh: No refresh token";
+    onRefreshError( QNetworkReply::AuthenticationRequiredError );
+    return;
+  }
+  if ( refreshTokenUrl_.isEmpty() )
+  {
+    qWarning() << "O2::refresh: Refresh token URL not set";
+    onRefreshError( QNetworkReply::AuthenticationRequiredError );
+    return;
+  }
+
+  QNetworkRequest refreshRequest( refreshTokenUrl_ );
+  refreshRequest.setHeader( QNetworkRequest::ContentTypeHeader, O2_MIME_TYPE_XFORM );
+  QMap<QString, QString> parameters;
+  parameters.insert( O2_OAUTH2_CLIENT_ID, clientId_ );
+  parameters.insert( O2_OAUTH2_CLIENT_SECRET, clientSecret_ );
+  parameters.insert( O2_OAUTH2_REFRESH_TOKEN, refreshToken() );
+  parameters.insert( O2_OAUTH2_GRANT_TYPE, O2_OAUTH2_REFRESH_TOKEN );
+
+  QByteArray data = buildRequestBody( parameters );
+  Q_ASSERT( mRefreshReply == nullptr );
+  mRefreshReply = getManager()->post( refreshRequest, data );
+
+  // Use direct connections contrary to O2::refresh()
+  connect( mRefreshReply, &QNetworkReply::finished, this, &QgsO2::onRefreshFinishedSynchronous, Qt::DirectConnection );
+  connect( mRefreshReply, qgis::overload< QNetworkReply::NetworkError >::of( &QNetworkReply::error ), this, &QgsO2::onRefreshErrorSynchronous, Qt::DirectConnection );
+
+  const int r_reqtimeout = oauth2config()->requestTimeout() * 1000;
+  QEventLoop rloop( nullptr );
```

Comment:
Hm -- could we use QgsBlockingNetworkRequest here instead of the event loop? I'm concerned that there's going to be times when the refresh is happening on the main thread, and using the event loop on the main thread will definitely cause issues...

Reply:
> could we use QgsBlockingNetworkRequest here instead of the event loop?

Good idea. I've just updated the commit to do that. Works well for the use case where I hit the issue. I've also tried with the layout composer and works fine here too.

Reply:
Thanks!

I still think in the long term we'll need to rethink the whole thread safety of the authentication system, but I'm happy for band aid fixes at the moment!